### PR TITLE
Astype keeps nan when converting into string

### LINKED ISSUE #25353: astype(str) / astype_unicode: np.nan converted to "nan" (checknull, skipna)
#### Code Sample

```python
>>> import pandas as pd
>>> import numpy as np
>>> pd.Series(["foo",np.nan]).astype(str)
```

#### Output
0    foo
1    nan # nan string
dtype: object

#### Expected output
0    foo
1    NaN # np.nan
dtype: object

#### Problem description

Upon converting this Series I would expect np.nan to remain np.nan but instead it is casted to string "nan". Maybe I'm alone in this and you'd actually expect that (I can't see a realistic use case for these string "nan" but well...). 
So I could figure out than upon using the code sample the Series' values are processed through astype_unicode in pandas._libs.lib.

There is a skipna argument in astype_unicode and I thought it would get passed along when using pd.Series.astype(str,skipna = True) but it does not. The docstring of pd.Series.astype does not mention skipna explicitely but mentions kwargs so I tried doing this while printing skipna in astype_unicode:

```
pd.Series(["foo",np.nan]).astype(str, skipna = True)
```

Skipna stayed to the default value False in astype_unicode so it does not get passed along.

However when using astype_unicode directly setting skipna to True will not change the output of the the code sample anyways because **checknull** does not seem to work properly.
You can test that by printing the result of checknull in lib.pyx as I did here:
https://github.com/ThibTrip/pandas/commit/4a5c8397304e3026456d864fd5aeb7b8b9adca5f

#### Input
```python
>>> from pandas._lib.lib import astype_unicode
>>> import numpy as np
>>> astype_unicode(np.array(["foo",np.nan]))
```
 #### Output
```
foo is null? False
nan is null? False
array(['foo', 'nan'], dtype=object)
```

#### Expected output
```
foo is null? False
nan is null? True
array(['foo', NaN], dtype=object)
```

I tried patching it (so not a proper fix) using the code below but it din't work.
Also would be nice to be able to do pd.Series.astype(str,skipna = True). Whether skipna should then be True or False as default is another matter.
```
>>> if not (skipna and checknull(arr_i)):
>>>     if arr_i is not np.nan:
>>>         arr_i = unicode(arr_i)
```

All of this was done in a developper version I installed today (see details below). The only alteration is the code in my commit linked above.
Sorry if this has been referenced before I searched in various ways and could not find anything except a similar issue with pd.read_excel and dtype str:

https://github.com/nikoskaragiannakis/pandas/commit/694849da2654d832d5717adabf3fe4e1d5489d43

Also very sorry for the mess with the commits I got a bit confused during my investigation (also I did not get enough sleep). Is it possible to delete all but my last commit? The other ones are irrelevant.

Cheers Thibault

<details>

INSTALLED VERSIONS
------------------
commit: 4a5c8397304e3026456d864fd5aeb7b8b9adca5f
python: 3.7.2.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.25.0.dev0+132.g4a5c83973.dirty
pytest: 4.2.0
pip: 19.0.1
setuptools: 40.7.3
Cython: 0.29.5
numpy: 1.15.4
scipy: 1.2.0
pyarrow: 0.11.1
xarray: 0.11.0
IPython: 7.2.0
sphinx: 1.8.4
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.9
feather: None
matplotlib: 3.0.2
openpyxl: 2.6.0
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml.etree: 4.3.1
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: 1.2.17
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: 0.2.0
fastparquet: 0.2.1
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by makbigc**
- closes #25353 
- 1 test added
- passes black pandas
- whatsnew entry

### CODE REVIEW

**Review by TomAugspurger — COMMENTED** [review]
This likely needs a deprecation cycle.

We also need to consider whether `skipna` is a valid keyword to `.astype` and add it to the function signature, document it, etc.

For now, I think a specific keyword for controlling how `np.nan` is converted with `.astype(str)` make the most sense. I'm not sure what the name would be.

**Comment by TomAugspurger** [comment]
Right, the skipna argument would only be necessary if we're deprecating.

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -621,9 +621,12 @@ def _astype(self, dtype, copy=False, errors="raise", **kwargs):
                 else:
                     values = self.get_values(dtype=dtype)
 
+                skipna = kwargs.pop("skipna", True)
                 # _astype_nansafe works fine with 1-d only
                 vals1d = values.ravel()
-                values = astype_nansafe(vals1d, dtype, copy=True, **kwargs)
+                values = astype_nansafe(
+                    vals1d, dtype, copy=True, skipna=skipna, **kwargs
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: not sure I understand why you changed this code, is this not already passed thru?

File: pandas/tests/frame/test_dtypes.py [hunk 2]
```diff
@@ -627,13 +627,13 @@ def test_astype_str(self):
         assert_frame_equal(result, expected)
 
     def test_astype_str_float(self):
-        # see gh-11302
+        # GH 25353
         result = DataFrame([np.NaN]).astype(str)
-        expected = DataFrame(["nan"])
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/tests/frame/test_dtypes.py]: IIRC wanted to match numpy

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.0.0.rst [hunk 3]
```diff
@@ -145,7 +145,7 @@ Indexing
 Missing
 ^^^^^^^
 
--
+- When converting into string, :meth:`Series.astype` will keep ``np.nan`` as missing value (:issue:`25353`)
```
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v1.0.0.rst]: yeah I think this is ok to actually make this change, but let's make it more prominent

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (makbigc) [reply pandas/core/internals/blocks.py]: `astype` method doesn't have `skipna` keyword. `skipna` is passed in `kwargs`.
https://github.com/pandas-dev/pandas/blob/e0c63b4cfaa821dfe310f4a8a1f84929ced5f5bd/pandas/core/generic.py#L5764

`skipna` is pop out from `kwargs` if any. If not, `skipna=True` is set and passed to `astype_nansafe`. (In `astype_nansafe`, `skipna=False` is default.) 

https://github.com/pandas-dev/pandas/blob/e0c63b4cfaa821dfe310f4a8a1f84929ced5f5bd/pandas/core/dtypes/cast.py#L663

So, the default behaviour after this change is to skip converting `nan` into string.

File: pandas/tests/series/test_dtypes.py [hunk 4]
```diff
@@ -147,7 +147,7 @@ def test_astype_datetime64tz(self):
     )
     def test_astype_str_map(self, dtype, series):
         # see gh-4405
-        result = series.astype(dtype)
+        result = series.astype(dtype, skipna=False)
```
[the remark this answers is not in the document]
Comment (makbigc) [reply pandas/tests/series/test_dtypes.py]: `astype` method doesn't have `skipna` keyword. `skipna` is passed in `kwargs`.
https://github.com/pandas-dev/pandas/blob/e0c63b4cfaa821dfe310f4a8a1f84929ced5f5bd/pandas/core/generic.py#L5764

`skipna` is pop out from `kwargs` if any. If not, `skipna=True` is set and passed to `astype_nansafe`. (In `astype_nansafe`, `skipna=False` is default.) 

https://github.com/pandas-dev/pandas/blob/e0c63b4cfaa821dfe310f4a8a1f84929ced5f5bd/pandas/core/dtypes/cast.py#L663

So, the default behaviour after this change is to skip converting `nan` into string.

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/tests/series/test_dtypes.py]: we shouldn't be passing non-accepted keywords in kwargs.

**Review by jreback — CHANGES_REQUESTED** [review]
need to revise the test where ``skipna`` is passed to ``.astype()`` which is not allowed. Can you create an issue about this (disallow kwargs in .astype).

File: doc/source/whatsnew/v1.0.0.rst [hunk 5]
```diff
@@ -39,6 +39,28 @@ Backwards incompatible API changes
...
 -
 
+.. _whatsnew_0250.api_breaking.series_string_conversion:
+
+String conversion of Series with nan
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+ When converting into string, :meth:`Series.astype` will not convert ``np.nan`` into string and keep it as missing value (:issue:`25353`)
+
+ *pandas 0.25.x*
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: use Previous behavior

File: doc/source/whatsnew/v1.0.0.rst [hunk 6]
```diff
@@ -39,6 +39,28 @@ Backwards incompatible API changes
...
+
+.. code-block:: ipython
+
+   In [1]: pd.Series(['foo', np.nan]).astype(str)
+   Out[2]:
+   0    foo
+   1    nan
+   dtype: object
+
+*pandas 1.0.0*
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: use New behavior

File: doc/source/whatsnew/v1.0.0.rst [hunk 7]
```diff
@@ -39,6 +39,28 @@ Backwards incompatible API changes
 - :class:`pandas.core.groupby.GroupBy.transform` now raises on invalid operation names (:issue:`27489`).
 -
 
+.. _whatsnew_0250.api_breaking.series_string_conversion:
+
+String conversion of Series with nan
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+ When converting into string, :meth:`Series.astype` will not convert ``np.nan`` into string and keep it as missing value (:issue:`25353`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: :meth:`Series.astype(str)` previously would coerce a ``np.nan`` to the string ``nan``. Now pandas will preserve the missing value indicator.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by makbigc** [comment]
> need to revise the test where `skipna` is passed to `.astype()` which is not allowed. Can you create an issue about this (disallow kwargs in .astype).

@jreback You said that `skipna` is non-accepted keyword in kwargs. But `skipna` is passed as kwargs in `astype` presently.

https://github.com/pandas-dev/pandas/blob/944e26ef81ff6fc4ca74f889ea442ae62357603d/pandas/core/generic.py#L5760

Is what you mean to change the API of `astype` as the following?

`    def astype(self, dtype, copy=True, errors="raise", skipna=True, **kwargs):`

But this modification would change the subsequent function's API? Do I misunderstand you?

File: doc/source/whatsnew/v1.0.0.rst [hunk 8]
```diff
@@ -39,6 +39,28 @@ Backwards incompatible API changes
 - :class:`pandas.core.groupby.GroupBy.transform` now raises on invalid operation names (:issue:`27489`).
 -
 
+.. _whatsnew_0250.api_breaking.series_string_conversion:
+
+String conversion of Series with nan
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:meth:Series.astype(str) previously would coerce a np.nan to the string nan. Now pandas will preserve the missing value indicator (:issue:`25353`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.0.rst]: IIUC, @jschendel and I both prefer a deprecation cycle. It looks like you're making a breaking change. Is that correct?

**Comment by rendorHaevyn** [comment]
Just so you're all aware, this non-feature also appears to impact NoneType:

```python
>>> na = np.array(['foo','bar',5,np.nan,None])
>>> [type(x) for x in na]
[<class 'str'>, <class 'str'>, <class 'int'>, <class 'float'>, <class 'NoneType'>]

>>> da = pd.Series(na)
>>> [type(x) for x in da.astype('str')]
[<class 'str'>, <class 'str'>, <class 'str'>, <class 'str'>, <class 'str'>]
```

**Comment by WillAyd** [comment]
Loosely thought about this after reading through #27949 but do we care to differentiate between the following like numpy?

```python
>>> np.array([1, np.nan, 3]).astype(object)
array([1.0, nan, 3.0], dtype=object)

>>> np.array([1, np.nan, 3]).astype(str)
array(['1.0', 'nan', '3.0'], dtype='<U32')
```

Coercing to an object in NumPy would maintain the actual NA value, but to a string would write as `'nan'`. I suppose this makes more sense in that world where object and string types are different, but I'm just wondering what kind of churn we would be causing if we actually did want `astype(str)` to stringify `nan` with a proper `StringArray` in place

**Comment by TomAugspurger** [comment]
I think that the `np.nan -> 'nan'` behavior is surprising to most people the first time they see it. So by default, I would want to exclude NA values from being astyped (after a deprecation cycle).

**Comment by makbigc** [comment]
Linting test failed (i.e., black formatting). But I can pass the black formatting check in my own machine.

**Comment by WillAyd** [comment]
@makbigc might want to give a pre-commit hook a try:

https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#python-pep8-black

In any case can you merge master and re-push?

**Review by jreback — COMMENTED** [review]
@makbigc couple of things here

- I don't think we want / need  a kwarg ``skipna`` in ``.astype()`` at all
- this is a pretty low-level change *just* in ``astype_nansafe`` itself.

we *also* want to deprecate (show a FutureWarning). So what this needs is a *just* a check if there are nans and if so then we need to deprecate it before just doing what we are doing now (converting to string).

**Comment by makbigc** [comment]
@jreback Anything else? Please tell me.

**Comment by jreback** [comment]
can you merge master and will look again

**Comment by makbigc** [comment]
The `-W error` is set for pytest command in azure-37-numpydev.

https://github.com/pandas-dev/pandas/blob/8806ed7120fed863b3cd7d3d5f377ec4c81739d0/ci/azure/posix.yml#L55-L59

A lot of tests were failed by the FutureWarning. What should I do?

**Comment by WillAyd** [comment]
@makbigc I think need to address https://github.com/pandas-dev/pandas/pull/28176#pullrequestreview-328622327 first; the problem now is that the warning is getting thrown all over the place because `skipna` was added to `astype`

If you can adjust design per comment can re-evaluate issues from there

**Comment by makbigc** [comment]
@WillAyd Thanks for your reply. The skipna parameter has been removed. The skipna is set True by default in `astype_nansafe`.

The tests are failed by the FutureWarning. https://github.com/pandas-dev/pandas/pull/28176#issuecomment-570247305

**Comment by makbigc** [comment]
@jreback The tests were failed by the issue of FutureWarning. https://github.com/pandas-dev/pandas/pull/28176#issuecomment-570247305

Any resolution?

**Comment by makbigc** [comment]
@jreback Any feedback? Thanks.

**Comment by languitar** [comment]
It seems this didn't make it into pandas 1.0 but also `skipna` has been removed from `astype`. What is the intended way of casting to string with NaNs in pandas 1.0?

**Comment by jorisvandenbossche** [comment]
If we are going to deprecate this, I think we need a keyword to opt in to the future behaviour, so you have the possibility to silence the warning. 

And if we are adding a keyword, it seems to make sense to use `skipna` for that, as that could (at least in 0.25) already be used for this purpose (although undocumented)

**Comment by TomAugspurger** [comment]
Where are we at here? Seems like the preference is for

```
>>> s = pd.Series([1, np.nan])
>>> type(s.astype(str)[1])
UserWarning("Converting NaN to string 'nan'. In the future NaN will remain as NaN. Specify skipna=True to silence this warning and adopt the new behavior, or skipna=False to restore the old behavior").
str

>>> type(s.astype(str, skipna=False)[1])  # no warning
str

>>> type(s.astype(str, skipna=True)[1])  # no warning
float
```

Is that right? If so, @makbigc are you interested in implementing that?

**Comment by simonjayhawkins** [comment]
some time has passed since this PR was opened and pandas 1.0 has since been released.

> If we are going to deprecate this, I think we need a keyword to opt in to the future behaviour, so you have the possibility to silence the warning.
> 
> And if we are adding a keyword, it seems to make sense to use `skipna` for that, as that could (at least in 0.25) already be used for this purpose (although undocumented)

So, for me, a deprecation cycle on old (incorrect) behaviour and keep the skipna keyword (but deprecate it) is probably now a requirement to keep a stable api through to pandas 2.0.

This approach addresses both the regression in #31708 and the unexpected behaviour of  #25353

**Comment by simonjayhawkins** [comment]
@makbigc can you merge master to resolve conflicts, fix-up failing tests and address https://github.com/pandas-dev/pandas/pull/28176#issuecomment-595887420

**Comment by simonjayhawkins** [comment]
@makbigc closing as stale. ping if you want to continue.

**Comment by NumberPiOso** [comment]
Hello, @simonjayhawkins, @tomausperge

I would like to continue this PR as it seems it would solve multiple issues, I just want to declare everything that must be done before and to ensure that this new behaviour is still preferred.

- [ ]  Add _skipna_  keyword default to False at `pandas/core/generic.NDFrame.astype` method. Following the UserWarning https://github.com/pandas-dev/pandas/pull/28176#issuecomment-595887420
- [ ] Document this new behavior.
- [ ] Start deprecation cycle

Open Issues related
- #25353 
- #28454
- #31708

Unmerged PRs related
- #35060

**Pull request closed without merging**